File: Python/Algorithm/Deleting_Edge.py
# Complexity:
# Work: Divides the tree into two parts so that product of the summation of the trees be maximum

def dfs(ver, par = 0):
    ssum[ver] += ver
    for child in tre[ver]:
        if child == par: continue
        dfs(child, ver)
        ssum[ver] += ssum[child]

# Reading the number of vertices
global tre, ssum
vr = int(input())
tre = [[] for _ in range(vr + 1)]
ssum = [0] * (vr + 1)

# Building the tree
for _ in range(vr - 1):
    v1, v2 = map(int, input().split())
    tre[v1].append(v2); tre[v2].append(v1)

# Start DFS from root (assuming root is 1)
dfs(1)

ans = 0
for i in range(2, vr + 1):
    part1 = ssum[i]; part2 = ssum[1] - part1
    ans = max(ans, (part1 * part2) % M)
print(ans)

File: CPP/Algorithm/Deleting_Edge.cpp
// Complexity:
// Work: Divides the tree into two parts so that product of the summation of the trees be maximum

vector<int64_t> tre[N]; int64_t ssum[N];
auto dfs(auto vr, auto par = 0) {
    ssum[vr] += vr;
    for (auto &child : tre[vr]) {
        if (child == par) continue;
        dfs(child, vr);
        ssum[vr] += ssum[child];
    }
}

// Operation:
    int64_t vr; cin >> vr;
    for (i = 0; i < vr - 1; ++i) {
        int64_t v1, v2; cin >> v1 >> v2;
        tre[v1].emplace_back(v2); tre[v2].emplace_back(v1);
    }
    dfs(rt); int64_t ans = 0;
    for (i = 2; i <= vr; ++i) {
        int64_t part1 = ssum[i], part2 = ssum[1] - part1;
        ans = max(ans, (part1 * part2) % M);
    }
    cout << ans;